src/analysis/processing/qgsalgorithmgrid.cpp

```diff
@@ -0,0 +1,527 @@
+  switch ( mIdx )
+  {
+    case 0: //point
+      createPointGrid(sink, feedback);
+      break;
+    case 1: //line
+      createLineGrid(sink, feedback);
+      break;
+    case 2: //rectangle
+      createRectangleGrid(sink, feedback);
+      break;
+    case 3: //diamond
+      createDiamondGrid(sink, feedback);
+      break;
+    case 4: //hexagon
+      createHexagonGrid(sink, feedback);
+      break;
+    default:
+      break; //no default needed
+  }
+
+
+  QVariantMap outputs;
+  outputs.insert( QStringLiteral( "OUTPUT" ), dest );
+  return outputs;
+}
+
+void QgsGridAlgorithm::createPointGrid(std::unique_ptr< QgsFeatureSink > &sink, QgsProcessingFeedback *feedback)
+{
+  QgsFeature f = QgsFeature();
+
+  int cols =  static_cast<int>( std::ceil(mGridExtent.width()) / (mHSpacing - mHOverlay) );
+  int rows = static_cast<int>( std::ceil(mGridExtent.height()) / (mVSpacing - mVOverlay) );
+
+  int id = 1;
+  int cnt = 0;
+  int cellcnt = rows * cols;
+  double cnt_log = static_cast<double>(cellcnt) * 0.01;
+
+  for(int col = 0; col < cols; col++)
+  {
+    for(int row = 0; row < rows; row++)
+    {
+      double x = mGridExtent.xMinimum() + (col * mHSpacing - col * mHOverlay);
+      double y = mGridExtent.yMaximum() - (row * mVSpacing - row * mVOverlay);
+
+      f.setGeometry(QgsGeometry().fromPointXY(QgsPoint( x, y )));
+      f.setAttributes(QgsAttributes() << id << x << y << x + mHSpacing << y + mVSpacing);
+      sink->addFeature(f, QgsFeatureSink::FastInsert);
+
+      id++;
+      cnt++;
+
+      if(std::fmod(static_cast<double>(cnt),cnt_log) == 0)
+      {
+        feedback->setProgress((static_cast<double>(cnt)/ static_cast<double>(cellcnt)) * 100 );
+      }
+      if ( feedback && feedback->isCanceled() )
+        break;
+    }
+  if ( feedback && feedback->isCanceled() )
+    break;
+  }
+}
+
+void QgsGridAlgorithm::createLineGrid( std::unique_ptr< QgsFeatureSink > &sink, QgsProcessingFeedback *feedback)
+{
+  QgsFeature f = QgsFeature();
+
+  QVector<double> hSpace;
+  if( mHOverlay > 0)
+  {
+    hSpace = QVector<double>()<< mHSpacing - mHOverlay << mHOverlay;
+  }
+  else
+  {
+    hSpace = QVector<double>()<< mHSpacing << mHSpacing;
+  }
+
+  QVector<double> vSpace;
+  if( mVOverlay > 0)
+  {
+    vSpace = QVector<double>()<< mVSpacing - mVOverlay << mVOverlay;
+  }
+  else
+  {
+    vSpace = QVector<double>()<< mVSpacing << mVSpacing;
+  }
+
+  int cnt = 0;
+  int id = 1;
+
+  //latitude lines
+  double cnt_max = mGridExtent.height() / mVSpacing;
+  double cnt_log = cnt_max * 0.01;
+  double y = mGridExtent.yMaximum();
+
+  while(y >= mGridExtent.yMinimum())
+  {
+    if( feedback && feedback->isCanceled() )
+      break;
+
+    QgsPoint pt1 = QgsPoint(mGridExtent.xMinimum(), y);
+    QgsPoint pt2 = QgsPoint(mGridExtent.xMaximum(), y);
+    QVector<QgsPoint> lineNodes = QVector<QgsPoint>() << pt1 << pt2;
+
+    QgsPolyline line = QgsPolyline(lineNodes);
+    f.setGeometry(QgsGeometry::fromPolyline(line));
```

Comment:
We can use an optimised line string constructor here (for raw speeeed!):
   
    f.setGeometry( QgsGeometry( new QgsLineString( pt1, pt2 ) ) );